ci: make minikube.sh work on macOS M1 with the qemu2 driver

zjx20 opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

This PR makes scripts/minikube.sh work on macOS M1 with the qemu2 driver.

To utilize minikube on macOS M1, the qemu2 VM driver is currently the only viable option. Therefore, the following modifications have been implemented:
* the RESOLV_CONF variable has been made configurable because it needs to be overridden as /etc/resolv.conf when using the qemu2 driver.
* adding additional disks when using the qemu2 driver.

Also, **the GNU sed should be used**.

Here is the usage example:

```bash
export KUBE_VERSION=v1.28.3    # it's the latest version supported by minikube v1.32.0.
export MINIKUBE_ARCH=arm64
export VM_DRIVER=qemu2         # hyperkit is unavailable for arm64, so qemu2 is the only option for macOS M1
export CNI=auto
export RESOLV_CONF=/etc/resolv.conf
export CPUS=4                  # "nproc" is unavailable on macOS, so explicitly set a number
export MEMORY=4096
export DISK_SIZE=20

./scripts/minikube.sh up
./scripts/minikube.sh deploy-rook
```

## Related issues ##

No.

## Future concerns ##

List items that are not part of the PR and do not impact it's
functionality, but are work items that can be taken up subsequently.

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release
  notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md)
  updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

riya-singhal31 commented:
Please update commit message type from _**scripts**_ to _**ci**_ and reduce the length of body's line to max 80 chars by splitting into a new line, to make the checks passed.

riya-singhal31 commented:
Have you tested this out?

zjx20 commented:
> Have you tested this out?

It works for my MBP M1 with macOS Monterey and minikube v1.32.0, but I haven't tested it on linux systems.

nixpanic commented:
/test ci/centos/k8s-e2e-external-storage/1.29

nixpanic commented:
> /test ci/centos/k8s-e2e-external-storage/1.29

Deploying in the CI still works :+1:

I can not test this on a Mac, ideally someone else tries it out before this gets merged.

zjx20 commented:
Thank you @nixpanic. I hope someone will kindly try this branch on macOS M1 and give some feedback.

Here are the instructions:

```bash
export KUBE_VERSION=v1.28.3    # it's the latest version supported by minikube v1.32.0.
export MINIKUBE_ARCH=arm64
export VM_DRIVER=qemu2         # hyperkit is unavailable for arm64, so qemu2 is the only option for macOS M1
export CNI=auto
export RESOLV_CONF=/etc/resolv.conf
export CPUS=4                  # "nproc" is unavailable on macOS, so explicitly set a number
export MEMORY=4096
export DISK_SIZE=20

./scripts/minikube.sh up
./scripts/minikube.sh deploy-rook
```

black-dragon74 commented:
@zjx20 Please use GNU sed on macOS. It will ensure compatibility of the script and will avoid the macOS specific changes (like adding extensions for inline replacements)

And then update this PR with `sed` specific changes removed?

Regards

zjx20 committed:
ci: make minikube.sh work on macOS M1 with the qemu2 driver

To utilize minikube on macOS M1, the qemu2 VM driver is currently the
only viable option. Therefore, the following modifications have been
implemented:
* the RESOLV_CONF variable has been made configurable because it needs
to be overridden as /etc/resolv.conf when using the qemu2 driver.
* adding additional disks when using the qemu2 driver.

Signed-off-by: zjx20 <zhoujianxiong2@gmail.com>

zjx20 commented:
> @zjx20 Please use GNU sed on macOS. It will ensure compatibility of the script and will avoid the macOS specific changes (like adding extensions for inline replacements)
> 
> And then update this PR with `sed` specific changes removed?
> 
> Regards

Hi @black-dragon74, I have reverted the changes related to the sed command and updated the PR description. Please take a look.

Madhu-1 commented:
/test ci/centos/k8s-e2e-external-storage/1.29

black-dragon74 commented:
> Hi @black-dragon74, I have reverted the changes related to the sed command and updated the PR description. Please take a look.

Thank you! Looks good to me.

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
@mergifyio queue
